File: weather_mpi.py
from operator import indexOf
import pandas as pd
import math
from pprint import pprint
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

if rank == 0:
    data_20 = pd.read_csv("data/2020_tor_data.csv")
    data_21 = pd.read_csv("data/2021_tor_data.csv")
    data_21.head()
    present_days = data_21[data_21["Month"] == 6]
    prev_days = data_20[data_20["Month"] == 6]    
    
    for i in range(size-1):
        comm.send(data_20, dest=i+1)
        comm.send(data_21, dest=i+1)
        comm.send(present_days, dest=i+1)
        comm.send(prev_days, dest=i+1)
else:
    start = MPI.Wtime()

    data_20 = comm.recv(source=0)
    data_21 = comm.recv(source=0)
    present_days = comm.recv(source=0)
    prev_days = comm.recv(source=0)
    
    present_days = present_days[(present_days["Day"] < 15 + rank) & (present_days["Day"] >= 8 + rank )]
    present_days = present_days[["Max Temp (°C)", "Min Temp (°C)", "Total Rain (mm)"]]
    CD = present_days.to_numpy()

    prev_days = prev_days[(prev_days["Day"] < 15 + rank) & (prev_days["Day"] >= 1 + rank)]
    prev_days = prev_days[["Max Temp (°C)", "Min Temp (°C)", "Total Rain (mm)"]]
    PD = prev_days.to_numpy()

    w_dct = {}

    for index in range(8):
        w_dct[f"W{index+1}"] = PD[index: index+7]

    ed_list = []
    # calculating means for present year 7 days
    present_max = present_min = present_rain = 0

    for line in CD:
        present_max += line[0]
        present_min += line[1]
        present_rain += line[2]

    present_max = present_max / 7
    present_min = present_min / 7
    present_rain = present_rain / 7

    # mean of parameters for each window
    for line in w_dct:
        max_sum = min_sum = rain_sum = 0
        for values in w_dct[line]:
            max_sum += values[0]
            min_sum += values[1]
            rain_sum += values[2]
        max_mean = max_sum / 7
        min_mean = min_sum / 7
        rain_mean = rain_sum / 7
        euclid_distance = math.sqrt((present_max - max_mean)**2 + (present_min - min_mean)**2 + (present_rain - rain_mean) ** 2)
        ed_list.append(euclid_distance)

    min_ed = min(ed_list)
    index_min_ed = indexOf(ed_list, min_ed)

    selected_prev_window = w_dct.get(f"W{index_min_ed+1}")

    
    present_variation_vector = []
    for i in range(len(CD[0])):
        vector = []
        for j in range(len(CD)-1):
            vector.append(CD[j+1][i] - CD[j][i])
        present_variation_vector.append(vector)

    prev_variation_vector = []
    for i in range(len(selected_prev_window[0])):
            vector = []
            for j in range(len(selected_prev_window)-1):
                vector.append(selected_prev_window[j+1][i] - selected_prev_window[j][i])
            prev_variation_vector.append(vector)
    
    mean_present_var = []
    for i in range(len(present_variation_vector)):
        mean_present_var.append(sum(present_variation_vector[i]) / len(present_variation_vector[i]))

    mean_prev_var = []
    for i in range(len(prev_variation_vector)):
        mean_prev_var.append(sum(prev_variation_vector[i]) / len(prev_variation_vector[i]))

    mean_variation_vector = []
    for i in range(len(mean_present_var)):
        mean_variation_vector.append((mean_present_var[i] + mean_prev_var[i]) / 2)
    
    prev_day = CD[-1]
    for i in range(len(prev_day)):
        prev_day[i] += mean_variation_vector[i]
    pred_arr = prev_day.tolist()

    pred_max = pred_arr[0]
    pred_min = pred_arr[1]
    pred_rain = pred_arr[2]

    actual_data = data_21[(data_21["Month"] == 6) & (data_21["Day"] == 15 + rank)]
    actual_values = actual_data[["Max Temp (°C)", "Min Temp (°C)", "Total Rain (mm)"]].to_numpy().flatten()

    print(f'''\nPrediction for June {15 + rank}, 2021 \n-----------------------------
\tMax Temp | Min Temp | Rainfall
Actual:   {actual_values[0]}   |   {actual_values[1]}   |   {actual_values[2]}
Predicted: {round(pred_max, 2)} |   {round(pred_min, 2)}  |   {round(pred_rain, 2)}
''')
    end = MPI.Wtime()
    print(f"Time taken to predict weather by process {rank}: {end}")

File: weather_mpi.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include "mpi.h"
#include "csv.h"
#include "utils.h"

char presentCSV[][50] = {
  "./data/2021_tor_data.csv",
  "./data/2021_van_data.csv",
  "./data/2021_mtl_data.csv",
  "./data/2021_cal_data.csv",
  "./data/2021_edm_data.csv",
  "./data/2021_ham_data.csv",
  "./data/2021_hfx_data.csv",
  "./data/2021_kel_data.csv",
  "./data/2021_kit_data.csv",
  "./data/2021_lon_data.csv",
  "./data/2021_osh_data.csv",
  "./data/2021_ott_data.csv",
  "./data/2021_peg_data.csv",
  "./data/2021_qbc_data.csv",
  "./data/2021_reg_data.csv",
  "./data/2021_sas_data.csv",
  "./data/2021_stcath_data.csv",
  "./data/2021_stjn_data.csv",
  "./data/2021_vic_data.csv",
  "./data/2021_win_data.csv"
};

char prevCSV[][50] = {
  "./data/2020_tor_data.csv",
  "./data/2020_van_data.csv",
  "./data/2020_mtl_data.csv",
  "./data/2020_cal_data.csv",
  "./data/2020_edm_data.csv",
  "./data/2020_ham_data.csv",
  "./data/2020_hfx_data.csv",
  "./data/2020_kel_data.csv",
  "./data/2020_kit_data.csv",
  "./data/2020_lon_data.csv",
  "./data/2020_osh_data.csv",
  "./data/2020_ott_data.csv",
  "./data/2020_peg_data.csv",
  "./data/2020_qbc_data.csv",
  "./data/2020_reg_data.csv",
  "./data/2020_sas_data.csv",
  "./data/2020_stcath_data.csv",
  "./data/2020_stjn_data.csv",
  "./data/2020_vic_data.csv",
  "./data/2020_win_data.csv" 
};

// Predicts the weather of a particular day of a year
void predictWeatherForGivenDay(
  double *prediction, 
  double *actual,
  char presentData[][50][50],
  char prevData[][50][50], 
  int day, 
  int *cols, 
  int numParams
) {
  // Get current data for the past 7 days
  char CD[7][50][50];
  filterMatrix(presentData, CD, day - 7, day - 1, cols, numParams);

  // Get previous data (1 year old) for a 14 day period
  char PD[14][50][50];
  filterMatrix(prevData, PD, day - 7, day + 6, cols, numParams);

  // 8 sliding windows based on previous data, 7 days per window
  double windows[8][7][numParams];
  memset(windows, 0, 8*7*numParams*sizeof(double));
  for (int i = 0; i < 8; i++) {
    for (int j = 0; j < 7; j++) {
      for (int k = 0; k < numParams; k++) {
        windows[i][j][k] = atof(PD[j+i][k]);
      }
    }
  }

  // Sum the present values 
  double presentSums[numParams];
  memset(presentSums, 0, numParams*sizeof(double));
  for (int i = 0; i < 7; i++)
    for (int j = 0; j < numParams; j++)
      presentSums[j] += atof(CD[i][j]);

  // Using the present sums, determine the means
  double presentMeans[numParams];
  memset(presentMeans, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++)
    presentMeans[i] = presentSums[i] / 7;

  // Determine the means of the previous values based on the
  // sliding windows and calculate the Euclidian Distances
  double edList[8];
  memset(edList, 0, 8*sizeof(double));
  for (int i = 0; i < 8; i++) {
    double prevSums[numParams];
    memset(prevSums, 0, numParams*sizeof(double));
    for (int j = 0; j < 7; j++)
      for (int k = 0; k < numParams; k++)
        prevSums[k] += windows[i][j][k];

    double prevMeans[numParams];
    memset(prevMeans, 0, numParams*sizeof(double));
    for (int j = 0; j < numParams; j++)
      prevMeans[j] = prevSums[j] / 7;

    int numPresentMeans = sizeof(presentMeans) / sizeof(presentMeans[0]);
    double euclidDistance = calcEuclidianDistance
    (
      presentMeans,
      prevMeans,
      numPresentMeans
    );

    edList[i] = euclidDistance;
  }

  // Determine the minimum Euclidian Distance and find its index
  double minEd = min(edList, sizeof(edList) / sizeof(edList[0]));
  int indexMinEd = 0;
  for (int i = 0; i < 8; i++) {
    if (edList[i] == minEd) {
      indexMinEd = i;
      break;
    }
  }

  // Isolate the window with the minimum Euclidian Distance
  double selectedPrevWindow[7][numParams];
  memset(selectedPrevWindow, 0, 7*numParams*sizeof(double));
  for (int i = 0; i < 7; i++)
    for (int j = 0; j < numParams; j++)
      selectedPrevWindow[i][j] = windows[indexMinEd][i][j];

  // Find variation vector based on present data
  double presentVariationVector[numParams][6];
  memset(presentVariationVector, 0, numParams*6*sizeof(double));
  for (int i = 0; i < numParams; i++)
    for (int j = 0; j < 6; j++)
      presentVariationVector[i][j] = atof(CD[j+1][i]) - atof(CD[j][i]);

  // Find variation vector based on previous data
  double prevVariationVector[numParams][6];
  memset(prevVariationVector, 0, numParams*6*sizeof(double));
  for (int i = 0; i < numParams; i++)
    for (int j = 0; j < 6; j++)
      prevVariationVector[i][j] = selectedPrevWindow[j+1][i] - selectedPrevWindow[j][i];

  // Determine the mean of the present variation vector
  double meanPresentVar[numParams];
  memset(meanPresentVar, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++) {
    double presentVarSum = 0;
    for (int j = 0; j < 6; j++)
      presentVarSum += presentVariationVector[i][j];
    meanPresentVar[i] = presentVarSum / 6;
  }

  // Determine the mean of the previous variation vector
  double meanPrevVar[numParams];
  memset(meanPrevVar, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++) {
    double prevVarSum = 0;
    for (int j = 0; j < 6; j++)
      prevVarSum += prevVariationVector[i][j];
    meanPrevVar[i] = prevVarSum / 6;
  }

  // Determine the mean of both present and previous mean variation vectors
  double meanVariationVector[numParams];
  memset(meanVariationVector, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++)
    meanVariationVector[i] = (meanPresentVar[i] + meanPrevVar[i]) / 2;

  // Retrieve values from the day previous to the day being predicted
  double previousDay[numParams];
  memset(previousDay, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++)
    previousDay[i] = atof(CD[6][i]);

  // Add mean variation vector values to the previous day to get final prediction
  for (int i = 0; i < numParams; i++)
    previousDay[i] += meanVariationVector[i];

  // Assign the final prediction values
  for (int i = 0; i < numParams; i++)
    prediction[i] = previousDay[i];

  // Retrive the actual values for comparision purposes
  char actualValues[1][50][50];
  filterMatrix(presentData, actualValues, day, day, cols, numParams);
  for (int i = 0; i < numParams; i++)
    actual[i] = atof(actualValues[0][i]);
}

int main(int argc, char **argv) {
  int rank, numProcs;
  double startTime = 0.0;
  double endTime = 0.0;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &numProcs);

  if (rank == 0)
    startTime = MPI_Wtime();

  // 9 - Max Temp (°C)
  // 11 - Min Temp (°C)
  // 13 - Mean Temp (°C)
  // 15 - Heat Deg Days (°C)
  // 17 - Cool Deg Days (°C)
  // 19 - Total Rain (mm)
  // 21 - Total Snow (cm)
  // 23 - Total Precip (mm)
  // 25 - Snow on Grnd (cm)
  int cols[] = {9, 11, 13, 15, 17, 19, 21, 23, 25};
  int numParams = sizeof(cols) / sizeof(cols[0]);
  int numLocations = sizeof(presentCSV) / sizeof(presentCSV[0]);

  for (int location = 0; location < numLocations; location++) {
    char presentData[400][50][50];
    char prevData[400][50][50];
    if (rank == 0) {
      // Get data from CSV files
      readCSV(presentData, presentCSV[location]);
      readCSV(prevData, prevCSV[location]);
    }

    // Retrieve data from rank 0
    MPI_Bcast(&presentData, 400*50*50, MPI_CHAR, 0, MPI_COMM_WORLD);
    MPI_Bcast(&prevData, 400*50*50, MPI_CHAR, 0, MPI_COMM_WORLD);

    // Divide days evenly between processes
    int count = 344 / numProcs;
    int remainder = 344 % numProcs;
    int startDay, endDay;
    if (rank < remainder) {
        startDay = rank * (count + 1);
        endDay = startDay + count;
    } else {
        startDay = rank * count + remainder;
        endDay = startDay + (count - 1);
    }

    // Add 8 since first 7 days cannot be predicted
    // printf("Rank %d starting from %d and ending at %d\n", rank, startDay+8, endDay+8);
    for (int i = startDay+8; i < endDay+8; i++) {
      double prediction[numParams];
      double actual[numParams];
      predictWeatherForGivenDay(
        prediction, 
        actual, 
        presentData, 
        prevData, 
        i,
        cols, 
        numParams
      );
      // Comment out these print statements for a more accurate time measurement
      // printf("\nFROM PROCESS %d Day: %d\nActual: ", rank, i);
      // for (int j = 0; j < numParams; j++) {
      //   printf("%.2lf ", actual[j]);
      // }
      // printf("\nPrediction: ");
      // for (int j = 0; j < numParams; j++) {
      //   printf("%.2lf ", prediction[j]);
      // }
      // printf("\n");
    }
  }

  MPI_Barrier(MPI_COMM_WORLD);

  if (rank == 0) {
    endTime = MPI_Wtime();
    double timeTaken = endTime - startTime;
    printf("\nTime taken: %f seconds\n", timeTaken);
  }

  MPI_Finalize();
  return 0;
}
